app/controllers/concerns/application_user_concern.rb

```diff
@@ -2,9 +2,10 @@
 module ApplicationUserConcern
   extend ActiveSupport::Concern
   include ApplicationUserMasqueradeConcern
+  include ApplicationUserOauthConcern
 
   included do
-    before_action :authenticate_user!, unless: :publicly_accessible?
+    before_action :authenticate!, unless: :publicly_accessible?
```

Comment:
Not sure if it can be done, but can we do something like `super` for `authenticate_user`?

Reply:
I don't see the need to call `authenticate_user!` at this point. The problem with `authenticate_user!` is that it either returns `User` from Warden, or immediately throws. I need to check the current user from Doorkeeper first, and only throw when both Devise and Doorkeeper do not have any users.

The only place, I can think of, to call `authenticate_user!` is in `authenticate!`. So, we'll have,

```diff
def current_user
  @current_user ||= current_user_from_devise || current_user_from_doorkeeper
end

def authenticate!
-  throw :warden unless devise_controller? || current_user
+  authenticate_user! unless current_user
end
```

Warden caches `User` so there won't be an issue of calling `warden.authenticate` multiple times. The only reason I can think of why one would do this is to just not `throw :warden` because it's Devise's internal implementation. But we are already breaking that when we use `warden.authenticate` in `current_user_from_devise` and `env['warden'].user` in `ApplicationCable::Connection.connect` anyway.

Reply:
You can let me know what do you think, and I'll defer to you.

Reply:
Ok then let's just keep it first